Feature/hide blank items - WIP

eric-kimbrel opened the pull request:
adds an option to hide blank items.  closes #55

msftclas commented:
@eric-kimbrel,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

darthtrevino commented:
@eric-kimbrel Did you mean to close this without merging the changes?

eric-kimbrel commented:
Yes, i found an issue i need to address.  When hiding blank items the value associated with that item still effects the scale used to render the other value bars which i think is not desired.   I'll get a fix for that worked up and open a new PR for it.

darthtrevino commented:
Ah, I see; it's fine either way in this case, but in the future you don't have to delete PR's that aren't ready - just slap a "(WIP)" to the end of the title.  We can use PRs like that as kind of a zone for discussion.

eric-kimbrel commented:
Ah, that makes sense.  I'll go ahead and reopen this one and mark it as WIP

msftclas commented:
@eric-kimbrel,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

eric-kimbrel commented:
Because the slicer expects data items to be handed to it with the item width percentages already calculated the filtering of blank categories has to be done on on the VIsual side rather than in the slicer itself.  I've removed the logic added to AttributeSlicer and updated AttributeSlicerVisual with the needed logic to filter data before passing it on.

darthtrevino commented on packages/attribute-slicer-powerbi/src/AttributeSlicerVisual.spec.ts:
```diff
@@ -382,6 +384,27 @@ describe("AttributeSlicerVisual", function () {
+        let update = createOptionsWithCategories(fakeCats, "SOME_CATEGORY_NAME");
+        instance.updateWithType(update, UpdateType.Data);
+
+        expect(attributeSlicer.data.map(function (n) { return n.match; })).to.be.deep.equal(fakeCats);
```
Rewrite as lambda: `n => n.match`

darthtrevino commented on packages/attribute-slicer-powerbi/src/AttributeSlicerVisual.spec.ts:
```diff
@@ -132,6 +133,7 @@ describe("AttributeSlicerVisual", function () {
                             Values: true,
                         },
                     }],
+                    objects: objects,
```
Use es6 object shorthand: `objects,`

darthtrevino commented on packages/attribute-slicer-powerbi/src/AttributeSlicerVisual.ts:
```diff
@@ -230,12 +229,14 @@ export default class AttributeSlicer extends VisualBase {
-                    // The colors have changed, so we need to reload data
-                    if (!oldState.colors.equals(newState.colors)) {
+                    // The colors or hide items setting have changed, so we need to reload data
+                    if (!oldState.colors.equals(newState.colors) || oldState.hideEmptyItems !== newState.hideEmptyItems) {
```
extract to predicate `isReloadRequired`
